src/System.Private.CoreLib/src/System/Runtime/InteropServices/ComActivator.cs

```diff
@@ -177,6 +198,72 @@ private static Type FindClassType(Guid clsid, string assemblyName, string typeNa
             throw new COMException(string.Empty, CLASS_E_CLASSNOTAVAILABLE);
         }
 
+        private static AssemblyLoadContext GetALC(string assemblyPath)
+        {
+            AssemblyLoadContext alc;
+            if (!s_AssemblyLoadContexts.TryGetValue(assemblyPath, out alc))
+            {
+                alc = AddNewALC(assemblyPath);
+            }
+
+            return alc;
+
+            AssemblyLoadContext AddNewALC(string path)
+            {
+                // The static collection of ALCs is readonly. If a new ALC entry needs to be added,
+                // then we must make a copy, add the new ALC, and update the static member.
+                lock (s_AssemblyLoadContextsLock)
+                {
+                    // Check the new collection prior to adding the new ALC if
+                    // the needed ALC was added in the interim.
+                    AssemblyLoadContext alcMaybe;
+                    if (!s_AssemblyLoadContexts.TryGetValue(path, out alcMaybe))
+                    {
+                        var newCollection = new Dictionary<string, AssemblyLoadContext>(s_AssemblyLoadContexts);
```

Comment:
You need to specify the comparer here again. Otherwise, it is going to use the default one.

Reply:
How much faster is a microbenchmark that creates COM objects because of this lock-free optimization?

Reply:
>You need to specify the comparer here again. Otherwise, it is going to use the default one.

That seems like a broken API for the "copy" of a `Dictionary`, but clearly has been this way forever. I never knew this semantic.

>How much faster is a microbenchmark that creates COM objects because of this lock-free optimization?

Unclear. Will get some numbers when I get in tomorrow - I don't have a work machine at home to run things.

Reply:
It is broken because the state of a dictionary is defined by 3 components - keys, values, and hash/compare. In this case by copying a dictionary all 3 must be copied or else construction of the new dictionary could fail for no obvious reason. This is especially true when a constructor accepts an interface that the class itself implements. No other other language that I can think of fails this basic copying concept. I see there is another constructor that accepts an `IEqualityComparer`, but that shouldn't be needed to "copy" a dictionary because passing in a `IDictionary` by default should not throw.

Reply:
I guess the definition of `IDictionary` is such that there is no implied concept of comparison since according to the doc it is merely a "Represents a generic collection of key/value pairs.". This is very misleading, but I guess technically consistent.

Reply:
> No other other language that I can think of fails this basic copying concept.

The method to create exact clone in .NET is typically called `Clone`. .NET does not have copy constructors (convention).

Reply:
> How much faster is a microbenchmark that creates COM objects because of this lock-free optimization?

No discernible impact for the basic scenario. The scenario where there would be contention with instantiating multiple ALCs is hard to measure, but we can defer the optimization until we hear from a customer.